Add initial spec

marcoscaceres opened the pull request:
Adds the initial spec based on [latest proposal](https://github.com/WICG/identity-credential/blob/main/digital-credentials-2-proposal.md).

closes #61


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***

### :boom: Error: 502 Bad Gateway :boom: ###

[PR Preview](https://github.com/tobie/pr-preview#pr-preview) failed to build. _(Last tried on Feb 3, 2024, 12:42 AM UTC)_.

<details>
<summary>More</summary>


PR Preview relies on a number of web services to run. There seems to be an issue with the following one:

:rotating_light: [Spec Generator](https://www.w3.org/2015/labs/) - Spec Generator is the web service used to build specs that rely on ReSpec.

:link: [Related URL](https://labs.w3.org/spec-generator/?type=respec&url=https%3A%2F%2Fraw.githubusercontent.com%2FWICG%2Fdigital-credentials%2F06d9a61d1e82545e5fcdf2d9e27daaddd080a78c%2Findex.html%3Fforce%3D1%26isPreview%3Dtrue)

```
error code: 502
```

_If you don't have enough information above to solve the error by yourself (or to understand to which web service the error is related to, if any), please [file an issue](https://github.com/tobie/pr-preview/issues/new?title=Error%20not%20surfaced%20properly&body=See%20WICG/digital-credentials%2357.)._
</details>

TallTed commented on index.html:
```diff
@@ -0,0 +1,293 @@
+        <p>
+          A credential provider is responsible for presenting the user with a
+          list of [=digital credential/request protocol=] specific providers
+          (e.g., a list of mobile drivers license or applications that hold
+          driver's licenses).
```
Is it `drivers license` or `driver's license`? 

In either case, these should be consistent throughout, which they currently aren't.

Sakurann replied:
18013-5 says `driving licence` (british english) :)

marcoscaceres replied:
Yeah, it should be "driver's license"... will fix throughout. 

Noting also that W3C specs are conical in  US-EN, but can be translated to other language variants (should we want to do a British English version).

samuelgoto commented on index.html:
```diff
@@ -0,0 +1,293 @@
+          company: "Apple Inc.",
+          companyURL: "https://apple.com",
+        },
+        // Add additional editors here.
```
Can we have someone from Chrome here (you can use me as a placeholder, we'll find someone to edit this more long term) so that we can make sure that what's in the spec also represents what we'd implement in Chrome.

Also, procedurally, can we agree on blocking normative PR merges with an LGTM from the two implementions?

It would be good if we could also get Ben from Mozilla here to help edit, in case Firefox has an immediate desire to implement this too.

samuelgoto replied:
Yeah, OWNERs could help here too.

Sakurann replied:
not sure how to parse `can we agree on blocking normative PR merges with an LGTM from the two implementions`. Do you intend to merge a PR if it has approvals from at least 2 implementors regardless of the number of people requesting for changes?

OR13 replied:
thats how I read it, you would just set implementation reps as code owners, and require approval from them.

marcoscaceres replied:
Yeah, basically. That would mirror (somewhat) the [WHATWG working mode](https://whatwg.org/working-mode) and well as the [WebAppsWG working mode](https://www.w3.org/2022/04/webapps-wg-charter.html#working-mode), both of which work really well. 

We can add OWNERS as that automatically flags folks for review, or we can easily set that up via the GitHub settings (that someone with write access needs to also approve it, and we basically just agree on who needs to give a ✅.... it's a lightweight process).

marcoscaceres replied:
Filed https://github.com/WICG/digital-credentials/issues/60 for OWNERS.

samuelgoto replied:
Re-opening this: didn't we agree to have co-editors (in addition to OWNERs) from other implementations on the last call?

samuelgoto commented on index.html:
```diff
@@ -0,0 +1,293 @@
+  <head>
+    <meta charset="utf-8">
+    <title>
+      Digital Credentials API
```
I personally think that the term Digital Credentials is the best that we could use, but I hear that this isn't something particularly shared by everyone. Maybe we could put up for discussion in a issue and rename it as a separate step?

tplooker replied:
Was there a reason to not use the "Identity Credential" terminology here?

tplooker replied:
Apologies I didn't see #55 had been raised.

samuelgoto commented on index.html:
```diff
@@ -0,0 +1,293 @@
+    "use strict";
+    // See https://github.com/w3c/respec/wiki/ for how to configure ReSpec
+    var respecConfig = {
+      github: "WICG/digital-credentials",
```
digital-credentails doesn't exist yet, does it?

marcoscaceres replied:
It does now :)

samuelgoto commented on index.html:
```diff
@@ -0,0 +1,293 @@
+  <body data-cite="credential-management-1 WebCryptoAPI">
+    <section id="abstract">
+      <p>
+        This document standardizes an API solution for requesting digital
```
```suggestion
        This document specifies an API solution for requesting digital
```

CGs can't create standards, right? We'd standardize this much later in the process when we move this spec into a WG?

samuelgoto commented on index.html:
```diff
@@ -0,0 +1,293 @@
+    <section id="abstract">
+      <p>
+        This document standardizes an API solution for requesting digital
+        identity credentials (e.g., a drivers license) or verifying some aspect
```
```suggestion
        identity credentials (e.g. a drivers license) or verifying some aspect
```

TallTed replied:
[`e.g.,` and `i.e.,` should always be followed by (and usually preceded by) a comma](https://www.dailywritingtips.com/comma-after-i-e-and-e-g/).

If you don't like those commas for whatever reason, the abbreviations should be replaced by text like `for example` or `that is to say`, which may or may not also call for following and/or preceding commas.

samuelgoto commented on index.html:
```diff
@@ -0,0 +1,293 @@
+      <p>
+        This document standardizes an API solution for requesting digital
+        identity credentials (e.g., a drivers license) or verifying some aspect
+        of a digital identity credential (e.g., proof of age). The API builds
```
```suggestion
        of a digital identity credential (e.g. proof of age). The API builds
```

samuelgoto commented on index.html:
```diff
@@ -0,0 +1,293 @@
+    </h2>
+    <pre class="idl">
+    partial interface CredentialsContainer {
+      Promise&lt;DigitalCredential&gt; requestIdentity(IdentityRequestOptions options);
```
```suggestion
      Promise&lt;DigitalCredential&gt; requestIdentity(DigitalCredentialRequestOptions options);
```

Maybe we should use the DigitalCredential prefix for these mas, because otherwise we are going to run very quickly into a collision with the Identity Credential maps defined [here](https://fedidcg.github.io/FedCM/). For example, the [IdentityCredentialRequestOptions](https://fedidcg.github.io/FedCM/#enumdef-identitycredentialrequestoptionscontext).

marcoscaceres replied:
Yeah, good idea.

samuelgoto commented on index.html:
```diff
@@ -0,0 +1,293 @@
+    </h2>
+    <pre class="idl">
+    partial interface CredentialsContainer {
+      Promise&lt;DigitalCredential&gt; requestIdentity(IdentityRequestOptions options);
```
We don't have to do this all at once, but I'm wondering if something like a `requestDigitalCredential` may be a better (more specific ) method name in comparison to `requestidentity` ... again, because this is going to collide with Chrome's implementation of the `IdentityCredential` which FedCM uses.

samuelgoto replied:
Point being: we don't have to rename this now in this PR, but wanted to give the heads up that Chrome ships IdentityCredential and we'd like to avoid the conflicts here.

marcoscaceres replied:
@samuelgoto,  filed https://github.com/WICG/digital-credentials/issues/61 so we don't lose it. Let's pick this up there.

samuelgoto replied:
Yeah, agreed that we don't want to block this initial PR on this. Let me resolve this to unblock it here, and we can follow up on the issue you create. Thanks!

samuelgoto commented on index.html:
```diff
@@ -0,0 +1,293 @@
+      requesting a digital identity credential.
+    </p>
+    <pre class="idl">
+    dictionary IdentityRequestOptions {
```
```suggestion
    dictionary DigitalCredentialRequestOptions {
```

As I said earlier, I'd use a `DigitalCredential` prefix as opposed to `Identity` prefix, because otherwise Chrome is going to quickly run into a conflict with the interfaces defined by the `IdentityCredential` maps, for example [this](https://fedidcg.github.io/FedCM/#dictdef-identitycredentialrequestoptions) one.

samuelgoto commented on index.html:
```diff
@@ -0,0 +1,293 @@
+    </p>
+    <pre class="idl">
+    [Exposed=Window, SecureContext]
+    interface DigitalCredential : Credential {
```
I think inheriting from `Credential` would create a problem for implementations, because, according to the Credential Manager spec, implementations MUST loop through the subclasses and look the internal methods.

https://www.w3.org/TR/credential-management-1/#credential-internal-methods

```
Unless otherwise specified, each [interface object](https://heycam.github.io/webidl/#dfn-interface-object) created for interfaces which [inherit](https://heycam.github.io/webidl/#dfn-inherit) from [Credential](https://www.w3.org/TR/credential-management-1/#credential) MUST provide implementations for at least one of these internal methods, overriding [Credential](https://www.w3.org/TR/credential-management-1/#credential)'s default implementations, as appropriate for the [credential](https://www.w3.org/TR/credential-management-1/#concept-credential) type. E.g., [§3.2 The PasswordCredential Interface](https://www.w3.org/TR/credential-management-1/#passwordcredential-interface), [§4.1 The FederatedCredential Interface](https://www.w3.org/TR/credential-management-1/#federatedcredential-interface), and [[WEBAUTHN]](https://www.w3.org/TR/credential-management-1/#biblio-webauthn).
```

We could emphasize on the "Unless otherwise specified" part, but the challenge here is that this `DigitalCredential` here really isn't a Credential Manager `Credential` in the sense that all of the other ones are (PublicKeyCredential, PasswordCredential, FederatedCredential and IdentityCredential).

So, I think we need to make a decision as to whether we want to integrate properly with the Credential Management API (e.g. implement the other methods, like `[[CollectFromCredentialStore]]`) or not (e.g. not subclass `Credential` here).

WDYT?

OR13 replied:
If these APIs end up very different, and the only difference in their names is the word `Digital` I think that will lead to a lot of bug reports and developer confusion.

Sakurann replied:
How much of the other methods of CredMan API do we think digital credentials API will reuse?
It has been pointed out number of times that "credential" in a conventional/CredMan refers to the things like password, wubauthn keypair, which is different from how "verifiable/digital credentials" refer to collection of attributes of the user attested by the issuer. It might be that despite all this, a lot of the same methods can be reused, I don't know, but that sounds like something that could influence the decision here.

marcoscaceres replied:
@samuelgoto, I agree that we should be building on the Credential Manager spec. The goal is be to eventually copy/paste this over to that spec. And yes, I agree that integration needs to be coherent and we are not doing a separate thing. 

At the same time, there was good reason for not using `.get()`: we can more strictly restrict the call site to just `DigitalCredentialRequestOptions` without allowing the mixing of credential types in the request (yes, we could also enforce that in `.get()`, but it's a bit more messy because it can't be enforced with WebIDL.... nevertheless, whatever we do spec should be `.get()` compatible). Also, calling `.requestIdentity()` was supposed to make it clear to developers which sheet/UI is going to be show: the one to pick a driver's license or similar credential, not for passwords or whatever other things are possible with straight up `.get()`. 

About who we should talk to, I think we should just talk to Credential Manager API folks... speaking as a ex-TAG member, I don't think the TAG will have much to offer here TBH (this is outside their purview, and can only provide very surface guidance... and will probably just tell us to go to talk to WebAppSec 🫠).

bvandersloot-mozilla replied:
Chiming in here that making this extend `Credential` would be something I believe we explicitly looked to avoid with the decision to use `.requestIdentity()`. If we subclass, that makes the more "login" centric `.get()` able to request an identity.

I think that the options we have are (1) making a parallel construction for driver's-license like things where we reuse algorithms as helpers or (2) making the `Credential` inherit a more generic object that we lift the shared bits into.

samuelgoto replied:
>  I think we should just talk to Credential Manager API folks

Yeah, I think that's the right next step.

> that makes the more "login" centric .get()

I think there are a couple of things that I'm struggling with this proposition:

1) First, I don't share the interpretation that the Credential Manager API is for "login".
2) Second, it is clear to me that the Digital Credentials here are going to be used for "login": it even shows up in regulation.

I think some of my confusion here is that i think maybe we don't quite understand what we mean by "login" (e.g. i think there is a big distinction between sign-up - i.e. account creation and recovery -, and sign-in).

Chrome expects FedCM to be used for a lot of [use cases](https://github.com/fedidcg/FedCM/blob/main/meetings/2021/Web%20Identity%20API.pdf) that I don't think I would call "login", so that's what I'm struggling to reconcile.

bvandersloot-mozilla replied:
By login I mean sign-in.  

> First, I don't share the interpretation that the Credential Manager API is for "login".

Sign-in is mentioned in the first sentence of the use cases section of that spec and the opening sentence of that spec is "Signing into websites is more difficult than it should be." 

> Second, it is clear to me that the Digital Credentials here are going to be used for "login": it even shows up in regulation.

I'm not a lawyer so I can't speak to what the EU is mandating, but is it something that we should encourage globally? I say not.

> Chrome expects FedCM to be used for a lot of [use cases](https://github.com/fedidcg/FedCM/blob/main/meetings/2021/Web%20Identity%20API.pdf) that I don't think I would call "login", so that's what I'm struggling to reconcile.

IMO the non-sign-in use cases covered by FedCM can be explained as the consequences of trying to sign in while simultaneously dealing with an out-of-the-page IDP in the best way possible.

samuelgoto replied:
Ok, I'm going to resolve this, since @bvandersloot-mozilla 's [alternative design](https://github.com/WICG/digital-credentials/issues/61#issuecomment-1921637318) really captures the essence of what I was struggling with, which is now reflected in this PR.

samuelgoto commented on index.html:
```diff
@@ -0,0 +1,293 @@
+      <tbody>
+        <tr>
+          <td>
+            `"mdoc"`
```
Can we add this section on a separate PR? Or, otherwise, add VC to this list? And if Safari is not interested in supporting VCs, than remove mDocs so that we can reserve the right to not support that either?

OR13 replied:
I agree with this comment, I would also suggest adding a citation to at least one digital credential format, that can be obtained at no cost, per: https://www.w3.org/Guide/process/tilt/normative-references.html

marcoscaceres replied:
I also agree. I'll take it out.

marcoscaceres replied:
Filed https://github.com/WICG/digital-credentials/issues/58 for us to figure the registry stuff out.

samuelgoto replied:
Thank you, resolving this here and following up on that issue. Thanks for kicking it off.

samuelgoto reviewed (commented):
Thanks for putting this together!! It is a great starting point and is really going to help us find the convergence that we need.

Comments inline!

samuelgoto commented on index.html:
```diff
@@ -0,0 +1,293 @@
+    dictionary IdentityRequestProvider {
+      required DOMString protocol;
+      required ArrayBuffer request;
+      required CryptoKey publicKey;
```
Can we remove this until we get some convergence on this:

https://github.com/WICG/identity-credential/issues/56

I, for one, am not convinced that we need to have `publicKey` here at all.

OR13 replied:
I think it can be valuable to have the public key here, but in the interest of landing the initial spec in github, I am supportive of seeing it added in a separate PR after this one.

hlozi replied:
> Can we remove this until we get some convergence on this:
> 
> #56
> 
> I, for one, am not convinced that we need to have `publicKey` here at all.

We discussed this in the  2023-12-18 call:

meeting notes: https://github.com/WICG/identity-credential/wiki/2023-12-18-Meeting-Notes#do-we-really-need-publickey-to-be-outside-of-the-request-property

Consensus on the following: ( copied from the issue) 

the public key property should be required and should also live at the top level (independent of protocol)
the value should be more structured than just a DOMString
maybe redundant with a key acquired via an out-of-band method, but cannot be contradictory
need to further refine HPKE, including crypto agility

Happy to re-discuss it if needed.

marcoscaceres replied:
@samuelgoto, can you live with this here for now? (we need to put the publicKey somewhere, so this seems like an ok start?)

I added a link to #49 below so we can continue to discuss.

samuelgoto replied:
It is not clear to me what the browser would (and should) do with the `publicKey`. Chrome implemented a prototype of this API deliberately delegating the validation of signatures to verifiers and holders, and it is not clear to me what benefit we get by pulling this responsibility to the browser. 

I can be convinced that interoperability/fragmentation is something that I think the browser can help, but I think it would be premature to over-generalize this aspect at the moment, since we don't even know what "protocols" are, let alone force all of them to comply to a specific way to check for digital signatures.

samuelgoto replied:
Resolving this, since you removed it in the last commit:

https://github.com/WICG/digital-credentials/pull/57#issuecomment-1911289914

OR13 commented on index.html:
```diff
@@ -0,0 +1,293 @@
+      shortName: "digital-credentials",
+      specStatus: "CG-DRAFT",
+      group: "wicg",
+      localBiblio: {
+        "ISO-18013-5": {
+          title: "ISO/IEC 18013-5:2021 Personal identification — ISO-compliant driving licence — Part 5: Mobile driving licence (mDL) application",
+          href: "https://www.iso.org/standard/69084.html",
+          publisher: "ISO",
+          date: "2021-09",
+          status: "Published",
+          seriesInfo: "ISO/IEC 18013-5:2021",
+        },
```
```suggestion
      localBiblio: {
        "ISO-18013-5": {
          title: "ISO/IEC 18013-5:2021 Personal identification — ISO-compliant driving licence — Part 5: Mobile driving licence (mDL) application",
          href: "https://www.iso.org/standard/69084.html",
          publisher: "ISO",
          date: "2021-09",
          status: "Published",
          seriesInfo: "ISO/IEC 18013-5:2021",
        },
        "vc-data-model-2.0": {
          title: "The Verifiable Credentials Data Model v2.0",
          href: "https://www.w3.org/TR/vc-data-model-2.0/",
          publisher: "W3C Verifiable Credentials Working Group",
          status: "WD"
        }
```

marcoscaceres replied:
@OR13, vc-data-model-2.0 is already in specref: https://www.specref.org/?q=vc-data-model-2.0

The reason I had to add "ISO-18013-5" to the local biblio is that it's not coming up in specref. I filed a bug to find out why: 
https://github.com/tobie/specref/issues/776

Would you mind dropping this suggestion.

marcoscaceres replied:
Resolving as not needed.

OR13 commented on index.html:
```diff
@@ -0,0 +1,293 @@
+        </tr>
+      </thead>
+      <tbody>
+        <tr>
+          <td>
+            `"mdoc"`
+          </td>
+          <td>
+            [[[ISO-18013-5]]]
+          </td>
+          <td>
+            A mobile driving license (mDL) request.
+          </td>
+        </tr>
```
```suggestion
        <tr>
          <td>
            `"mdoc"`
          </td>
          <td>
            [[[ISO-18013-5]]]
          </td>
          <td>
            A mobile document (mDL) request.
          </td>
        </tr>
        <tr>
          <td>
            `"vc"`
          </td>
          <td>
            [[[vc-data-model-2.0]]]
          </td>
          <td>
            A verifiable credential (vc) request.
          </td>
        </tr>
```

OR13 replied:
actually based on https://github.com/WICG/identity-credential/blob/main/digital-credentials-2-proposal.md#the-identityrequestoptions-dictionary I think perhaps this should be `oid4vp` ?

marcoscaceres replied:
Taking all this out... let's move to https://github.com/WICG/digital-credentials/issues/58

OR13 commented on index.html:
```diff
@@ -0,0 +1,293 @@
+            [[[ISO-18013-5]]]
+          </td>
+          <td>
+            A mobile driving license (mDL) request.
```
does this work for other document types, or just mobile drivers licenses?

marcoscaceres replied:
I was just starting with this because it will be a miracle if we even get driver licenses.

marcoscaceres replied:
Note: removing this and moving discussion to https://github.com/WICG/digital-credentials/issues/58

awoie reviewed (changes requested):
It would be good to have IETF SD-JWT VC as well.

awoie reviewed (changes requested):
VCDM 2.0 is not a protocol. It technically is agnostic to securing mechanisms. IMO, this should be more specific.

Sakurann reviewed (changes requested):
Should also include IETF SD-JWT VC and not just ISO mdoc and W3C VCDM

tplooker commented on index.html:
```diff
@@ -0,0 +1,293 @@
+        credential is issued by an issuer (e.g., a government agency).
+      </dd>
+      <dt>
+        <dfn>Identity credential provider</dfn>
```
Just noting here the use of "identity credential provider" when the top level term is "digital credential", I think these terms should be consistent.

marcoscaceres replied:
Yeah, I kinda also like the generic "credential provider"... let's start with that. 

Eventually we should have a little diagram showing the model.

marcoscaceres replied:
https://github.com/WICG/digital-credentials/issues/66

tplooker commented on index.html:
```diff
@@ -0,0 +1,293 @@
+      <li>For each |provider:IdentityRequestProvider| of |options|'s
+      {{IdentityRequestOptions/providers}}.
+        <ol>
+          <li>[=verify the public key=] of |provider|.
```
Do we mean verify here in the sense of a cryptographic operation, or do we mean validate? e.g, make sure it is well formed and the type is expected/supported?

tplooker replied:
If we mean verify can we clarify what is being verified?

OR13 replied:
I assumed something like cert check here, I imagined something like x5c in jwk.

marcoscaceres replied:
We would be verifying that a `CryptoKey` is of the right type. 

A crypto key has the following properties: 

```WebIDL
enum KeyType { "public", "private", "secret" };

enum KeyUsage { "encrypt", "decrypt", "sign", "verify", "deriveKey", "deriveBits", "wrapKey", "unwrapKey" };

[SecureContext,Exposed=(Window,Worker)]
interface CryptoKey {
  readonly attribute KeyType type;
  readonly attribute boolean extractable;
  readonly attribute object algorithm;
  readonly attribute object usages;
};
```

So we would be **verifying** that the `type` is  "public", and reject "private" and "secret" keys... and then continue verifying whatever else we can from the above things.

marcoscaceres replied:
Let's discuss here https://github.com/WICG/digital-credentials/issues/56

Sakurann commented on index.html:
```diff
@@ -0,0 +1,293 @@
+      <dd>
+        The request protocol is a standardized format/structure to [=request
+        digital identity credentials=]. For example, a request protocol could
+        be a "mDoc" for mobile drivers licenses as per [[?ISO-18013-5]].
```
```suggestion
        be a "mdoc" for mobile drivers licenses as per [[?ISO-18013-5]].
```

18013-5 terminology is `mdoc` all lowercase i think

marcoscaceres replied:
Just noting that I'm removing this... was just showing what it could look like. 

@Sakurann, looking forward to discussing in https://github.com/WICG/digital-credentials/issues/58

bvandersloot-mozilla commented on index.html:
```diff
@@ -0,0 +1,308 @@
+    </h2>
+    <pre class="idl">
+    partial interface CredentialsContainer {
+      Promise&lt;DigitalCredential&gt; requestIdentity(IdentityRequestOptions options);
```
Reiterating from the call: It may be useful to abstract the DigitalCredential return type here and design an algorithm that both FedCM and DigitalCredential can hook into to provide identities. That way FedCM can intermingle in UI with 3-party ids and also with WebAuthN, but mDLs and WebAuthN can't be requested interchangably, which was the motivating concern for this API shape IIRC.

marcoscaceres replied:
I haven't read the FedCM spec yet... but yeah, so long as this can be a generic thing without a normative requirement on FedCM, then that works. 

I'm imagining `DigitalCredential` (or whatever we end up calling it) will end up in the Credential Management spec eventually.

bvandersloot-mozilla replied:
That sounds perfect. That just means FedCM would subclass DigitalCredential, in the same way it currently subclasses Credential and would put itself in the registry.

I agree that the generic `DigitalCredential` (or whatever) would live well in the Credential Management spec.

samuelgoto replied:
Can we try to agree with the goals here first? 

> both FedCM and DigitalCredential can hook into to provide identities

This matches much of my intuition, and it is what Chrome would like to see this design converge to. 

@marcoscaceres, does this match your intuition?

marcoscaceres replied:
I don't have a strong opinion (because I'm still not very familiar with FedCM). If both specs can use the same thing, then great. However, the problem remains about `.get()` conflating requesting multiple credential types, and it's impossible for an implementation to distinguish (because they are all part of the same `CredentialCreationOptions` dictionary).

In the [Extensions Points](https://www.w3.org/TR/credential-management-1/#implementation-extension) section, the CM spec encourages spec authors to define "a reasonable mechanism for dealing with requesting **credentials of distinct types**", which is what `. requestIdentity()` is trying to do, and then to pitch that to the Web App Sec WG. 

I think we need to document our options here and present our options to the Web App Sec WG to help us decide what the right course of action is here. 

I think we have general agreement that this poses issues:

```JS
const opts = {  publicKey , identity,  providers };
// who wins? what happens? What UI gets shown? 
navigator.credentials.get(opts);
``` 

Which, again, is the core of the problem. 

The `. requestIdentity()` disambiguates the problem by providing a unique method signature, only allowing this new `DigitalCredentialRequestOptions` dictionary, which deliberately does not inherit from `CredentialRequestOptions` to avoid its limitations, while providing the same functionality.

marcoscaceres replied:
Just as tl;dr: let's move this discussion to https://github.com/WICG/digital-credentials/issues/61

samuelgoto replied:
Ok, I think we arrived at a good conclusion on https://github.com/WICG/digital-credentials/issues/61. This issue here is now resolved from where I stand, and since we are following @bvandersloot-mozilla 's suggestion, I think it would be consistent with his thoughts. @marcoscaceres is there anything else that you think is actionable here?

marcoscaceres replied:
I'm happy to proceed. Let's do additional discussions in a follow up.

marcoscaceres commented:
Thanks again everyone for the amazing and super timely feedback 🙏 

Please check that I have addressed your feedback and please give an explicit ✅ if you think this is enough for us to start with. 

Remember, nothing is set in stone here... this is just an unofficial strawperson and, to paraphrase what @samuelgoto said during the call: we assume the whole thing is wrong and likely will need a few rewrites till we get something stable. 

Please file issues and we can go from there!

marcoscaceres commented:
I've dropped publicKey for now until we sort out when/how it will be used... maybe we don't need it? cc @samuelgoto

marcoscaceres commented:
This PR is currently blocked on resolving #61. 

Next steps are to discuss the solutions in #61 and set up a call with the editors of the Credential Management Spec. If folks have other opinions or solutions you'd like to discuss, please raise them there.

OR13 commented on index.html:
```diff
@@ -0,0 +1,310 @@
+      specStatus: "CG-DRAFT",
+      group: "wicg",
+      localBiblio: {
+        "ISO-18013-5": {
+          title: "ISO/IEC 18013-5:2021 Personal identification — ISO-compliant driving licence — Part 5: Mobile driving licence (mDL) application",
+          href: "https://www.iso.org/standard/69084.html",
+          publisher: "ISO",
+          date: "2021-09",
+          status: "Published",
+          seriesInfo: "ISO/IEC 18013-5:2021",
+        },
```
```suggestion
```

OR13 replied:
Currently unused, so not harm in removing... if the goal is to single intended supported digital credential formats, I suggest starting with a full set or starting with an empty set.

samuelgoto replied:
+1 for the "full set or empty set" framing.

marcoscaceres commented on index.html:
```diff
@@ -0,0 +1,299 @@
+    </h3>
+    <aside class="issue" data-number="65"></aside>
+    <h2>
+      The `DigitalIdentityRequestOptions` dictionary
```
```suggestion
       Extensions to `CredentialRequestOptions` dictionary
```

marcoscaceres commented on index.html:
```diff
@@ -0,0 +1,299 @@
+      The `DigitalIdentityRequestOptions` dictionary
+    </h2>
+    <pre class="idl">
+    dictionary DigitalIdentityRequestOptions : CredentialRequestOptions {
```
```suggestion
    partial dictionary CredentialRequestOptions {
```

marcoscaceres commented on index.html:
```diff
@@ -0,0 +1,299 @@
+    </h2>
+    <pre class="idl">
+    dictionary DigitalIdentityRequestOptions : CredentialRequestOptions {
+      required sequence&lt;IdentityRequestProvider&gt; providers;
```
```suggestion
     sequence&lt;IdentityRequestProvider&gt; providers;
```

marcoscaceres commented on index.html:
```diff
@@ -0,0 +1,299 @@
+      The `providers` member
+    </h3>
+    <p>
+      The <dfn data-dfn-for="DigitalIdentityRequestOptions">providers</dfn>
```
```suggestion
      The <dfn data-dfn-for="CredentialRequestOptions">providers</dfn>
```

samuelgoto commented on index.html:
```diff
@@ -0,0 +1,302 @@
+    <section id="abstract">
+      <p>
+        This document specifies an API to allow user agents to mediate access
+        to, and representing, a digital identity (e.g. a digital driver's
```
```suggestion
        to, and representing, a verifiable issued digital identity (e.g. a government issued digital driver's license, a passport issued by a country, a student card or diploma issued by a university, an employee card issued by a company, a passenger's boarding pass issued by an airline, a profile issued by an online social network, membership cards, vaccination records, etc)  
```

samuelgoto replied:
I think it would be useful to make sure we are all on the same page with regards to how we are expecting this API to be used, so I added a few more concrete examples that I hear.

Does this match your intuition?

TallTed replied:
Whether or not the suggestion above is adopted, it is unclear what `and representing` is connected to.
* Does it mean `to allow user agents to represent a digital identity`?
* Does it mean `an API representing a digital identity`?
* Does it mean something else?

Once this is answered, I will draft a rephrasing ... unless such a rephrasing is how the answer is provided.

----

Separately, I suggest changing `verifiable issued digital identity` to `verifiably-issued digital identity` or `verifiable digital identity`, unless it is really meant to say that an _issued digital identity_ is _verifiable_, which makes me want a clear definition of how an _issued digital identity_ differs from an _**un**issued digital identity_

marcoscaceres replied:
It would be good to include these here now, but eventually move the expanded examples into the yet to be written introduction. 

Also agree with @TallTed, should be "verifiably-issued digital identity". I'll change your suggestion to that. 

@TallTed, can you file a bug instead for us to work on the abstract, intron, and scope?

marcoscaceres replied:
@TallTed filed https://github.com/WICG/digital-credentials/issues/68

samuelgoto replied:
> Also agree with @TallTed, should be "verifiably-issued digital identity". I'll change your suggestion to that.
"Verifiably-issued digital identity" works for me for now.

samuelgoto commented on index.html:
```diff
@@ -0,0 +1,302 @@
+    };
+    </script>
+  </head>
+  <body data-cite="credential-management-1 WebCryptoAPI">
```
```suggestion
  <body>
```

Bad copy/pasting?

marcoscaceres replied:
This one is ok (except we don't need Web Crypto). We cite "credential-management-1" in various places, so this lets Respec know the context.

samuelgoto replied:
Yep, my suggestion was just to remove the "WebCryptoAPI" here.

samuelgoto commented on index.html:
```diff
@@ -0,0 +1,302 @@
+  <head>
+    <meta charset="utf-8">
+    <title>
+      Digital Identities
```
```suggestion
      Digital Identity API
```

Maybe more consistent with other spec names? That is, it is an API specifically, rather than an overview of digital identities broadly speaking?

marcoscaceres replied:
We've generally moved away from having "API" in the names of specs (e.g., "Credential Management", "Permissions", "Fetch", etc.) to cover both the generalized concepts and the API.

samuelgoto replied:
Ha, didn't realize that was the case. I'll update a few more of my specs that I maintain to follow that then.

Resolving.

samuelgoto commented on index.html:
```diff
@@ -0,0 +1,302 @@
+    </h2>
+    <dl class="definitions" data-sort="">
+      <dt>
+        <dfn>Identity credential</dfn>
```
```suggestion
        <dfn>Identity Credential</dfn>
```

marcoscaceres replied:
I've prefer we keep "Credential" lower-case, just for consistency.

samuelgoto replied:
non-camel-case sgtm

samuelgoto commented on index.html:
```diff
@@ -0,0 +1,302 @@
+      <dd>
+        <p>
+          A specialized type of [=credential=] corresponding to the real-world
+          identity of a person enabling developers to make authentication
```
```suggestion
          identity of a person enabling a [=Verifier=] to make authentication
```

marcoscaceres replied:
we probably don't need to upper case concepts, as they are linked (and we can alternatively do that with CSS if we want to make them more salient).

samuelgoto commented on index.html:
```diff
@@ -0,0 +1,302 @@
+        <p>
+          A specialized type of [=credential=] corresponding to the real-world
+          identity of a person enabling developers to make authentication
+          decisions based on verified identity information.
```
```suggestion
          decisions based on verified identity statements made by an [=Issuer=].
```

TallTed replied:
This distinction is important. Verifiable Credentials do not provide any indication of the veracity of the statements they contain; they only offer verification that the Issuer made those statements. 
```suggestion
          decisions based on identity statements verifiably made by an [=Issuer=].
```

marcoscaceres replied:
Good point, @TallTed.

samuelgoto commented on index.html:
```diff
@@ -0,0 +1,302 @@
+      TBW
+    </p>
+    <h2>
+      Model
```
We are operating at the browser level, so our definition of "Credential" (in the Credential Manager sense) is different than other group's definition of "Credential", but there are many other concepts that I think are well articulated by the ISO and Verifiable Credentials community.

[Here](https://www.w3.org/TR/vc-data-model/) is one example of a data model that I think has a lot of good definitions.

I'm wondering if we are going to get held back agreeing on these definitions, and if we are better off just having the WebIDLs in this initial PR to unblock implementation ... WDYT?

Happy to go either way, but my intuition is that the WebIDL without the definitions more minimally viable.

marcoscaceres replied:
Let's run with this for now but let's definitely start taking definitions/concepts from other specs if we can. However, we should control whatever we feel strongly fall squarely in the purview of our spec.

samuelgoto commented on index.html:
```diff
@@ -0,0 +1,302 @@
+    <pre class="idl">
+    dictionary IdentityRequestProvider {
+      required DOMString protocol;
+      required ArrayBuffer request;
```
I'm also wondering if `request` should be an `object` too, but I think we can leave that for a subsequent PR.

bc-pi replied:
I suspect most of the folks working on "credentials" from IETF/OIDF communities are envisioning the protocol-specific request content as a string. OpenID4VP is all string based, for example, so whatever incarnation of that in conjunction with this API eventually emerges would be most naturally expressed as a string. I think I share @samuelgoto's view here.

samuelgoto replied:
@marcoscaceres resolved this and changed to DOMString in this version of the spec. Resolving this.

samuelgoto reviewed (commented):
This PR really captured what I think is a design that will reconcile many different aspects of the space: reconciling the two party model with the three party model, reconciling the distinction between identity verification and sign-in, reusing the Credential Management spec and algorithms, etc.

I made some comments to try to clarify what I think is already shared understanding. 

The backbone and the structure of this PR is solid. I'm playing with the code to see what would be involved in implementing, but my intuition is that this would set us in the right long term direction.

@bvandersloot-mozilla WDYT?

samuelgoto commented on index.html:
```diff
@@ -0,0 +1,302 @@
+    </h2>
+    <pre class="idl">
+    partial dictionary CredentialRequestOptions {
+      sequence&lt;IdentityRequestProvider&gt; providers;
```
So, something that i think is a bit inconsistent with the CredentailsContainer but I think is actually good, is that the CredMan spec goes over in the [Credential Registry](https://w3c.github.io/webappsec-credential-management/#sctn-cred-type-registry) the [options member identifier](https://w3c.github.io/webappsec-credential-management/#credential-type-registry-options-member-identifier), which is an WebIDL attribute that tells you which credential type to use.

```js
await navigator.credentials.get({
  publicKey: /** give me a passkey */,
  otp: /** give me an SMS OTP */,
  identity: /** give me a federated assertion */,
});
```

Whereas `providers` here, would be for `DigitalIdentity`, but that seems like a general term that could apply to other `Identity` subclasses.

```js
await navigator.identity.get({
  providers: [] // what Identity interface do we expect to provide?
});
```

For example, if we introduced a `FederatedIdentity` in the future, then we wouldn't know whether "digital" identities or "federated" identities are being requested.

So, this is a bit different from the `navigator.credentials.get()` API, but I'm not sure if it is necessarily bad.

It is hard to entirely predict the future, but I think the biggest differentiation in the subclasses of `Identity` is actually going to come as a result of introducing new `protocols`. That is, I think that the only distinction between a `decentralized` identity and a `federated` identity is the mechanism that defines how to request them.

That is, if/when we'd like to expose FedCM into this API, we could introduce a new `protocol`, add it to the registry in this spec here, and continuing returning a `DigitalIdentity` or, if necessary, create a subclass of `Identity` that is a `FederatedIdentity`, and the way we would figure out which one to construct is based on `protocol`, NOT the [options member identifier](https://w3c.github.io/webappsec-credential-management/#credential-type-registry-options-member-identifier)  as used in `navigator.credentials.get()`.

I think this could work, but since it deviates from `navigator.credentials.get()` it seemed worth noting.

@bvandersloot-mozilla @marcoscaceres WDYT?

marcoscaceres replied:
We should maybe just note this issue and say that `providers` may change based on what you describe above.

samuelgoto replied:
I'm going to resolve this here to unblock this draft, so that we can resolve asynchronously here in this issue:

https://github.com/WICG/digital-credentials/issues/65#issuecomment-1924941840

samuelgoto commented on index.html:
```diff
@@ -0,0 +1,302 @@
+        license). The API builds on [[[credential-management-1]]] as a means to
+        request a digital identity from the user agent or underlying platform.
+      </p>
+    </section>
```
Maybe a big example up front would help explain?

```suggestion
<div class=example>
Example showing how a website can request a digital identity.

```javascript
let {protocol, data} = await navigator.identity.get({
  signal: controller,
  mediation: "optional", 
  providers: [{
    protocol: "fake-protocol",
    request: "{type: 'drivers-license'}",
    nonce: "0xCAFE",
  }]
});
```
</div>

    </section>
```

marcoscaceres replied:
Definitely want to add examples, but I think we should hold till we have the first item in the registry. Then we can add real functional examples.

marcoscaceres replied:
Filed https://github.com/WICG/digital-credentials/issues/69

samuelgoto replied:
Sure, that works too.

samuelgoto commented on index.html:
```diff
@@ -0,0 +1,302 @@
+  <head>
+    <meta charset="utf-8">
+    <title>
+      Digital Identities
```
I personally feel that "digital identities" are inclusive of "federated identities", but just wanted to make sure that that's the case, because, if not, then, maybe we'd need to use a more general term (i.e. remove the "digital" prefix)?

My intuition is that "digital identities" is inclusive and a good starting point, but just wanted to be transparent of how I'd want to see this spec evolve over time.

marcoscaceres replied:
Tbh, I think we might go through a few more renames over the next few weeks.

samuelgoto replied:
Agreed. I think Digital Identities is a great starting point. Resolving.

marcoscaceres commented on index.html:
```diff
@@ -0,0 +1,302 @@
+      The following items are out of scope:
+    </p>
+    <ul>
+      <li>Enrollment process for establishing a [=digital identity=].
```
basically, what I mean is: "I go to the DMV (or equivalent in whatever country) and I get issued a physical and digital driver's license... somehow, the driver's license ends up in my digital wallet". 

I agree that that will eventually be in scope. Just trying to keep the problem space very constrained for now.

samuelgoto replied:
Resolving this, since we can have this discussion elsewhere in an issue.

marcoscaceres commented on index.html:
```diff
@@ -0,0 +1,302 @@
+    <pre class="idl">
+    dictionary IdentityRequestProvider {
+      required DOMString protocol;
+      required ArrayBuffer request;
```
```suggestion
      required DOMString request;
```

marcoscaceres commented on index.html:
```diff
@@ -0,0 +1,302 @@
+    </p>
+    <p>
+      The {{IdentityRequestProvider/protocol}} member's value is be one of the
+      keys defined in [[[#protocol-registry]]].
```
Let's keep protocol as a DOMString because we are going to move the Registry to another document (which will make updating the enum painful once this is on the W3C Rec Track). We can add a note that it should be implemented as an enum, and mimic enum behavior via prose. We do the same thing in the Permission spec.

samuelgoto replied:
SGTM

marcoscaceres commented on index.html:
```diff
@@ -0,0 +1,302 @@
+      The following items are out of scope:
+    </p>
+    <ul>
+      <li>Enrollment process for establishing a [=digital identity=].
```
```suggestion
      <li> Issuance process for establishing a [=digital identity=].
```

marcoscaceres commented on index.html:
```diff
@@ -0,0 +1,302 @@
+      The following items are out of scope:
+    </p>
+    <ul>
+      <li> Issuance process for establishing a [=digital identity=].
```
```suggestion
      <li>Issuance process for establishing a [=digital identity=].
```

samuelgoto commented on index.html:
```diff
@@ -0,0 +1,310 @@
+    [Exposed=Window, SecureContext]
+    interface DigitalIdentity : Identity {
+      readonly attribute DOMString protocol;
+      readonly attribute ArrayBuffer data;
```
```suggestion
      readonly attribute DOMString data;
```

samuelgoto reviewed (approved):
LGTM

There is a lot missing in this spec right now, but it contains a great baseline that we can work from!

Thanks for putting this together!

RByers commented:
retroactive LGTM as well, I like this direction. Thanks Marcos and Sam!

bc-pi commented:
Is the "rendered" version of this available and if so where? (sorry if that's an ignorant question - I am largely unfamiliar with w3c tooling and stuff)

samuelgoto commented:
https://wicg.github.io/digital-credentials/

bc-pi commented:
Thanks @samuelgoto !

marcoscaceres commented:
Renamed it to https://wicg.github.io/digital-identities/ ... to match the new spec title.
